Fix example code, clean up doc, and attempt to address customer feedback (#27189)
dotnet/docs: docs/standard/security/walkthrough-creating-a-cryptographic-application.md

---
title: "Walkthrough: Create a Cryptographic Application"
description: Walk through the creation of a cryptographic application. Learn how to encrypt and decrypt content in a Windows Forms application.
ms.date: 11/29/2021
dev_langs:
  - "csharp"
  - "vb"
helpviewer_keywords:
  - "cryptography [NET], example"
  - "cryptography [NET], cryptographic application example"
  - "cryptography [NET], application example"
ms.topic: tutorial
---

# Walkthrough: Create a Cryptographic Application

> [!NOTE]
> This article applies to Windows.
>
> For information about ASP.NET Core, see [ASP.NET Core Data Protection](/aspnet/core/security/data-protection/introduction).

> This walkthrough demonstrates how to encrypt and decrypt the contents a file. The code examples are designed for a Windows Forms application. This application does not demonstrate real-world scenarios, such as using smart cards. Instead, it demonstrates the fundamentals of encryption and decryption.

Reviewer: ```suggestion
This walkthrough demonstrates how to encrypt and decrypt the contents of a file. The code examples are designed for a Windows Forms application. This application does not demonstrate real-world scenarios, such as using smart cards. Instead, it demonstrates the fundamentals of encryption and decryption.
```

This walkthrough uses the following guidelines for encryption:

- Use the <xref:System.Security.Cryptography.Aes> class, a symmetric algorithm, to encrypt and decrypt data by using its automatically generated <xref:System.Security.Cryptography.SymmetricAlgorithm.Key%2A> and <xref:System.Security.Cryptography.SymmetricAlgorithm.IV%2A>.
- Use the <xref:System.Security.Cryptography.RSA> asymmetric algorithm to encrypt and decrypt the key to the data encrypted by <xref:System.Security.Cryptography.Aes>. Asymmetric algorithms are best used for smaller amounts of data, such as a key.

  > [!NOTE]
  > If you want to protect data on your computer instead of exchanging encrypted content with other people, consider using the <xref:System.Security.Cryptography.ProtectedData> class.

The following table summarizes the cryptographic tasks in this topic.

| Task | Description |
|--|--|

> | Creating a Windows Forms application | Lists the controls that are required to run the application. |

Reviewer: ```suggestion
| Create a Windows Forms application | Lists the controls that are required to run the application. |
```

> | Declaring global objects | Declares string path variables, the <xref:System.Security.Cryptography.CspParameters>, and the <xref:System.Security.Cryptography.RSACryptoServiceProvider> to have global context of the <xref:System.Windows.Forms.Form> class. |

Reviewer: ```suggestion
| Declaring global objects | Declares string path variables, the <xref:System.Security.Cryptography.CspParameters>, and the <xref:System.Security.Cryptography.RSACryptoServiceProvider> to have global context of the <xref:System.Windows.Forms.Form> class. |
```
```suggestion
| Declare global objects | Declares string path variables, the <xref:System.Security.Cryptography.CspParameters>, and the <xref:System.Security.Cryptography.RSACryptoServiceProvider> to have global context of the <xref:System.Windows.Forms.Form> class. |
```

> | Creating an asymmetric key | Creates an asymmetric public and private key-value pair and assigns it a key container name. |

Reviewer: ```suggestion
| Create an asymmetric key | Creates an asymmetric public and private key-value pair and assigns it a key container name. |
```

> | Encrypting a file | Displays a dialog box to select a file for encryption and encrypts the file. |
> | Decrypting a file | Displays a dialog box to select an encrypted file for decryption and decrypts the file. |
> | Getting a private key | Gets the full key pair using the key container name. |
> | Exporting a public key | Saves the key to an XML file with only public parameters. |
> | Importing a public key | Loads the key from an XML file into the key container. |
> | Testing the application | Lists procedures for testing this application. |

Reviewer: ```suggestion
| Encrypt a file | Displays a dialog box to select a file for encryption and encrypts the file. |
| Decrypt a file | Displays a dialog box to select an encrypted file for decryption and decrypts the file. |
| Get a private key | Gets the full key pair using the key container name. |
| Export a public key | Saves the key to an XML file with only public parameters. |
| Import a public key | Loads the key from an XML file into the key container. |
| Test the application | Lists procedures for testing this application. |
```

|--------------------------------------------|--------------------------|---------------------------|
| <xref:System.Windows.Forms.Button>         | `buttonEncryptFile`      | Encrypt File              |
| <xref:System.Windows.Forms.Button>         | `buttonDecryptFile`      | Decrypt File              |
| <xref:System.Windows.Forms.Button>         | `buttonCreateAsmKeys`    | Create Keys               |
| <xref:System.Windows.Forms.Button>         | `buttonExportPublicKey`  | Export Public Key         |
| <xref:System.Windows.Forms.Button>         | `buttonImportPublicKey`  | Import Public Key         |
| <xref:System.Windows.Forms.Button>         | `buttonGetPrivateKey`    | Get Private Key           |
| <xref:System.Windows.Forms.Label>          | `label1`                 | Key not set               |
| <xref:System.Windows.Forms.OpenFileDialog> | `_encryptOpenFileDialog` |                           |
| <xref:System.Windows.Forms.OpenFileDialog> | `_decryptOpenFileDialog` |                           |

Double-click the buttons in the Visual Studio designer to create their event handlers.

> ## Declaring global objects

Reviewer: ```suggestion
## Declare global objects
```

Add the following code as part of the declaration of the class Form1. Edit the string variables for your environment and preferences.

[!code-csharp[CryptoWalkThru#1](../../../samples/snippets/csharp/VS_Snippets_CLR/CryptoWalkThru/cs/Form1.cs#1)]
[!code-vb[CryptoWalkThru#1](../../../samples/snippets/visualbasic/VS_Snippets_CLR/CryptoWalkThru/vb/Form1.vb#1)]

> ## Creating an asymmetric key

Reviewer: ```suggestion
## Create an asymmetric key
```

This task creates an asymmetric key that encrypts and decrypts the <xref:System.Security.Cryptography.Aes> key. This key was used to encrypt the content and it displays the key container name on the label control.

Add the following code as the `Click` event handler for the `Create Keys` button (`buttonCreateAsmKeys_Click`).

[!code-csharp[CryptoWalkThru#2](../../../samples/snippets/csharp/VS_Snippets_CLR/CryptoWalkThru/cs/Form1.cs#2)]
[!code-vb[CryptoWalkThru#2](../../../samples/snippets/visualbasic/VS_Snippets_CLR/CryptoWalkThru/vb/Form1.vb#2)]

> ## Encrypting a file

Reviewer: ```suggestion
## Encrypt a file
```
